File: 113. Path Sum II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {    // Recursive DFS with information from upper layers
    const result = [];
    (function dfs(node, target, ancestors) {
        if (node) {
            const newAncestors = ancestors.concat([node.val]);
            if (node.left) {
                dfs(node.left, target - node.val, newAncestors);
            }
            if (node.val === target && !node.left && !node.right) {
                result.push(newAncestors);
            }
            if (node.right) {
                dfs(node.right, target - node.val, newAncestors);
            }
        }
    })(root, targetSum, []);
    return result;
};

File: Golang/Tree/113.Path-sum-ii.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, targetSum int) [][]int {
	ret := [][]int{}
	type TreeNodeWithPath struct {
		Node *TreeNode
		Path []int
	}
	var dfs func(node *TreeNodeWithPath, target int)
	dfs = func(node *TreeNodeWithPath, target int) {
		newPath := append(append([]int{}, node.Path...), node.Node.Val)
		if node.Node.Val == target && node.Node.Left == nil && node.Node.Right == nil {
			ret = append(ret, newPath)
		}
		if node.Node.Left != nil {
			dfs(&TreeNodeWithPath{
				Node: node.Node.Left,
				Path: newPath,
			}, target-node.Node.Val)
		}
		if node.Node.Right != nil {
			dfs(&TreeNodeWithPath{
				Node: node.Node.Right,
				Path: newPath,
			}, target-node.Node.Val)
		}
	}
	if root != nil {
		dfs(&TreeNodeWithPath{
			Node: root,
			Path: []int{},
		}, targetSum)
	}
	return ret
}
